src/GafferUI/Gadget.cpp

```diff
@@ -56,6 +56,54 @@ using namespace std;
 
 GAFFER_GRAPHCOMPONENT_DEFINE_TYPE( Gadget );
 
+//////////////////////////////////////////////////////////////////////////
+// Gadget::Signals
+//
+// We allocate these lazily because they have a significant overhead
+// in both memory and construction time, and for many Gadget
+// instances they are never actually used.
+//////////////////////////////////////////////////////////////////////////
+
+struct Gadget::Signals : boost::noncopyable
+{
+
+	VisibilityChangedSignal visibilityChangedSignal;
+	RenderRequestSignal renderRequestSignal;
+
+	ButtonSignal buttonPressSignal;
+	ButtonSignal buttonReleaseSignal;
+	ButtonSignal buttonDoubleClickSignal;
+	ButtonSignal wheelSignal;
+
+	EnterLeaveSignal enterSignal;
+	EnterLeaveSignal leaveSignal;
+	ButtonSignal mouseMoveSignal;
+
+	DragBeginSignal dragBeginSignal;
+	DragDropSignal dragEnterSignal;
+	DragDropSignal dragMoveSignal;
+	DragDropSignal dragLeaveSignal;
+	DragDropSignal dragEndSignal;
+	DragDropSignal dropSignal;
+
+	KeySignal keyPressSignal;
+	KeySignal keyReleaseSignal;
+
+	// Utility to emit a signal if it has been created, but do nothing
```

Comment:
The wording feels a little misleading here, technically it emits a signal if _any_ signal has been created right? Is there in any merit in actually lazily creating only the used signals, or do people tend to connect to most of them if they connect to one?

Reply:
> technically it emits a signal if any signal has been created

This way of putting it is even more misleading - it implies you can create signals individually.

> Is there in any merit in actually lazily creating only the used signals, or do people tend to connect to most of them if they connect to one?

There could be, depending on what you're optimising for, and depending on the access patterns. I expect there are patterns whereby only the press/drag signals are accessed for some Gadgets, so that is a possible sub-grouping. But each sub-grouping would require another pointer member in Gadget, and I was pleasantly surprised with the improvement in draw times merely from reducing the Gadget footprint.

I suppose you wouldn't necessarily need a second pointer if instead `Signals` had pointers to sub-groups of its own. But that can be done at a later date without affecting compatibility. Right now I'm happy enough with the improvements here - they're pretty much guaranteed to be a worthwhile improvement in all cases, whereas attempts at sub-grouping would be a bit more of a balancing act where more real-world data would be needed.

Reply:
Cool cool. LGTMT